pkg/icingadb/cleanup.go

```diff
@@ -0,0 +1,124 @@
+package icingadb
+
+import (
+	"context"
+	"fmt"
+	"github.com/icinga/icingadb/internal"
+	"github.com/icinga/icingadb/pkg/com"
+	"github.com/icinga/icingadb/pkg/driver"
+	"github.com/icinga/icingadb/pkg/types"
+	"github.com/icinga/icingadb/pkg/utils"
+	"golang.org/x/sync/errgroup"
+	"time"
+)
+
+// CleanupResult is the result of executing a cleanup routine and
+// stores how many rows were deleted and how many statements were executed for that.
+type CleanupResult struct {
+	Count  uint64
+	Rounds uint64
+}
+
+// CleanupStmt defines information needed to compose cleanup statements.
+type CleanupStmt struct {
+	Table string
+	PK    string
+	Where string
+}
+
+// Build assembles the cleanup statement for the specified database driver with the given limit.
+func (stmt *CleanupStmt) Build(driverName string, limit uint) string {
+	switch driverName {
+	case driver.MYSQL:
+		return fmt.Sprintf(`DELETE FROM %s WHERE %s LIMIT %d`, stmt.Table, stmt.Where, limit)
+	case driver.PGSQL:
+		return fmt.Sprintf(`WITH rows AS (
+SELECT %[1]s FROM %[2]s WHERE %[3]s LIMIT %[4]d
+)
+DELETE FROM %[2]s WHERE %[1]s IN (SELECT %[1]s FROM rows)`, stmt.PK, stmt.Table, stmt.Where, limit)
+	default:
+		panic("invalid database type")
+	}
+}
+
+// CleanupOlderThan deletes all rows with the specified statement that are older than the given time.
+// Deletes a maximum of as many rows per round as defined in count.
+// If the query execution time exceeds the duration defined in the grace period,
+// the count is halved and the query is rerun.
+// The count is reset once the query has run within the grace period.
```

Comment:
Wouldn't this, if the database is never able to delete more than `limit` rows per `grace`, result in always trying a useless query in between?

Reply:
Removed the grace period (see above).